### ISSUE TITLE
Enable Fast Forward Merge between Staging and Master

### ISSUE THREAD

**Issue by CMCDragonkai** [MatrixAI/TypeScript-Demo-Lib-Native#14]
### Specification

Right now the automatic merge between staging and master occurs through a merge commit.

Having a merge commit in master is not a good idea because it means we have to remember to rebase our staging onto master each time after merge.

Github doesn't support native fast forward merges. https://github.com/github/feedback/discussions/4618

Therefore it has to be done manually within the Gitlab CI/CD.

This means the `integration:merge` job rather than just triggering auto-merge, it will need to do a proper `git merge --ff-only staging`.

```sh
git checkout master
git merge --ff-only staging
git push upstream master
```

The only issue right now is that we would have to ensure we check out the `master`. Which means using `git remote add -f upstream ...`.

Note that gitlab's origin points elsewhere.

Now when doing the `git push` it will need authentication. I'd prefer not to use SSH keys for this usage, and instead an auth token that is capable of the push.

I believe we should be able to achieve this with our `GITHUB_TOKEN` being used for HTTP authentication.

Will need to understand how git does HTTP authentication.

### Additional context

* https://github.com/github/feedback/discussions/4618

### Tasks

1. Find out about HTTP authentication for `git`
2. Try using it to do a `git push`
3. Make sure you can do a `git checkout upstream/master` in the CI/CD
4. Change the `integration:merge` job to do this, the PR can still be created, it will just be automatically merged once the `integration:merge` job is pushed up

Note that if we are doing this, perhaps we can also just do tagging directly... and push up those changes, it's kind of complicated since it is concurrent, and those tag commits may be in conflict with new changes. So we leave that till later.

**Comment by CMCDragonkai** [comment]
Using something like this?

```
git push https://<GITHUB_ACCESS_TOKEN>@github.com/<GITHUB_USERNAME>/<REPOSITORY_NAME>.git
```

Or:

```
git remote set-url origin https://scuzzlebuzzle:<MYTOKEN>@github.com/scuzzlebuzzle/ol3-1.git
```

I don't think the username is necessary, the token should be sufficient.

**Comment by CMCDragonkai** [comment]
It's dangerous to leave this to manual rebasing, even I forgot that the tag ended up pointing to nowhere cause the rebase occurred after the tagging of `npm version`.

**Comment by CMCDragonkai** [comment]
I believe fast forward merge should be working now.

However the branch protection for commit status must be removed. This is because GH is blocking merges to master until the pipeline finishes. And in this case, the pipeline hasn't finished, while we are doing a `git push upstream HEAD:master`.

Or the the integration job will need to send in a commit status. Currently the `gh` command doesn't a way of submitting a commit status check. It would require using a `curl` call. For now let's not bother with this.

In doing this I learned a few things about how remotes work.

The fast forward merge required:

1. The `GIT_DEPTH: 0` to ensure that GL CICD would fetch all commits for the the ref spec, which is basically the staging branch that we are on.
2. `git remote add upstream "https://${GITHUB_TOKEN}@github.com/..."` is sufficient
3. Using `git checkout origin/master` results in a detached head, as in a local branch isn't created.
4. Using `git merge --ff-only origin/staging` will only do fast forward merge if possible, the history must be compatible
5. Finally `git push upstream HEAD:master`, the `HEAD:master` is what allows one to be in the detached head to proceed. The detached HEAD is where the fast forward merge occurred.

In the case of `gh pr create`, the usage of the `--body ""` option meant that we weren't relying on the command to figure out what to write in the body. If we used `--fill` instead, additional requirements are needed.

```
git remote add upstream ...
git fetch upstream staging master
git checkout -B staging upstream/staging
gh pr create ...
```

Basically the `--fill` option requires the source and target branches of the upstream remote to be fetched, AND you need to be on a local branch branch tracking the source upstream branch.

This is alot of requirements just to compute the relevant commit messages, so I'll leave this for posterity but won't implement it.

**Comment by CMCDragonkai** [comment]
Furthermore it does make sense to change the default branch to `staging`.

It is very easy to make a mistake and get new feature branches from `master`.

**Comment by CMCDragonkai** [comment]
As a precaution, `master` now does have a limitation being:

![image](https://user-images.githubusercontent.com/640797/169961442-5bf35f9d-2931-427d-bc80-e65d5d209dd2.png)

That is linear history is required. This should prevent any one clicking on the github PR to merge as that would produce a merge commit.
